compiler/src/iree/compiler/Codegen/Common/test/block_dynamic_dims.mlir

```diff
@@ -0,0 +1,101 @@
+  flow.dispatch.tensor.store %result, %output_in, offsets = [0, 0, 0, 0], sizes = [4, %m, 32, 128], strides = [1, 1, 1, 1]
+      : tensor<4x?x32x128xf16> -> !flow.dispatch.tensor<writeonly:tensor<4x?x32x128xf16>>{%m}
+  return
+}
+// CHECK-LABEL: func @block_attention_dims()
+//   CHECK-DAG:   %[[C32:.+]] = arith.constant 32 : index
+//   CHECK-DAG:   %[[C16:.+]] = arith.constant 16 : index
+//   CHECK-DAG:   %[[M:.+]] = flow.dispatch.workload.ordinal %{{.+}}, 0 : index
+//   CHECK-DAG:   %[[K2:.+]] = flow.dispatch.workload.ordinal %{{.+}}, 1 : index
+//   CHECK-DAG:   %[[M_DYNAMIC:.+]] = arith.divui %[[M]], %[[C16]]
+//       CHECK:   %[[Q_BINDING:.+]] = hal.interface.binding.subspan
+//  CHECK-SAME:       binding(0)
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x?x16x32x128xf16>>{%[[M_DYNAMIC]]}
+//       CHECK:   %[[K2_DYNAMIC:.+]] = arith.divui %[[K2]], %[[C32]]
+//       CHECK:   %[[K_BINDING:.+]] = hal.interface.binding.subspan
+//  CHECK-SAME:       binding(1)
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x?x32x32x128xf16>>{%[[K2_DYNAMIC]]}
+//       CHECK:   %[[V_BINDING:.+]] = hal.interface.binding.subspan
+//  CHECK-SAME:       binding(2)
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x?x32x32x128xf16>>{%[[K2_DYNAMIC]]}
+//       CHECK:   %[[MASK_BINDING:.+]] = hal.interface.binding.subspan
+//  CHECK-SAME:       binding(3)
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x32x?x16x?x32xf16>>{%[[M_DYNAMIC]], %[[K2_DYNAMIC]]}
+//       CHECK:   %[[OUTPUT_BINDING:.+]] = hal.interface.binding.subspan
+//  CHECK-SAME:       binding(4)
+//  CHECK-SAME:       !flow.dispatch.tensor<writeonly:tensor<4x?x16x32x128xf16>>{%[[M_DYNAMIC]]}
+//       CHECK:   %[[Q:.+]] = flow.dispatch.tensor.load %[[Q_BINDING]]
+//  CHECK-SAME:       sizes = [4, %[[M_DYNAMIC]], 16, 32, 128]
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x?x16x32x128xf16>>{%[[M_DYNAMIC]]}
+//       CHECK:   %[[K:.+]] = flow.dispatch.tensor.load %[[K_BINDING]]
+//  CHECK-SAME:       sizes = [4, %[[K2_DYNAMIC]], 32, 32, 128]
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x?x32x32x128xf16>>{%[[K2_DYNAMIC]]}
+//       CHECK:   %[[V:.+]] = flow.dispatch.tensor.load %[[V_BINDING]]
+//  CHECK-SAME:       sizes = [4, %[[K2_DYNAMIC]], 32, 32, 128]
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x?x32x32x128xf16>>{%[[K2_DYNAMIC]]}
+//       CHECK:   %[[MASK:.+]] = flow.dispatch.tensor.load %[[MASK_BINDING]]
+//  CHECK-SAME:       sizes = [4, 32, %[[M_DYNAMIC]], 16, %[[K2_DYNAMIC]], 32]
+//  CHECK-SAME:       !flow.dispatch.tensor<readonly:tensor<4x32x?x16x?x32xf16>>{%[[M_DYNAMIC]], %[[K2_DYNAMIC]]}
+//       CHECK:   %[[ATTENTION:.+]] = iree_linalg_ext.attention
+//       CHECK:       ins(%[[Q]], %[[K]], %[[V]], %{{.+}}, %[[MASK]] :
+//       CHECK:   %[[GENERIC:.+]] = linalg.generic
+//       CHECK:   flow.dispatch.tensor.store %[[GENERIC]], %[[OUTPUT_BINDING]]
```

Comment:
It would be nice to have a couple smaller test cases, maybe
```
func {
  tensor.empty()
  any dps (linalg) op
}
func {
  %0 = flow.dispatch.tensor.load
  %1 = flow.dispatch.tensor.load
  linalg ins(%0) outs(%1)
}
```
i.e. unit tests for empty and flow.dispatch.tensor.load so we don't have one mega test case to look at.

Reply:
I havent made the pass work for any linalg operation yet. It is just for attention ops. So I can remove the bindings and make these function arguments, but I really want to test that the expand_shape/collapse_shapes fold away.

Reply:
(boarding a plane so will be slow to respond) yeah not asking to remove this test, just add some smaller unit ones as well, but I see now that it didn't support directly expanding linalg ops